components/team_sidebar/team_sidebar.tsx

```diff
@@ -2,44 +2,339 @@
 // See LICENSE.txt for license information.
 
 import React from 'react';
+import Scrollbars from 'react-custom-scrollbars';
+import {FormattedMessage} from 'react-intl';
 
-import {Team} from 'mattermost-redux/types/teams';
+import classNames from 'classnames';
 
-import TeamButton from './team_button';
+import {DragDropContext, Droppable, DroppableProvided, DropResult} from 'react-beautiful-dnd';
 
-type Props = {
-    teams: Team[];
-};
+import {Dispatch} from 'redux';
+
+import Permissions from 'mattermost-redux/constants/permissions';
+
+import {Team, TeamMembership} from 'mattermost-redux/types/teams';
+import {Dictionary} from 'mattermost-redux/types/utilities';
+
+import {GenericAction, GetStateFunc} from 'mattermost-redux/types/actions';
+
+import {Constants} from 'utils/constants.jsx';
+import {filterAndSortTeamsByDisplayName} from 'utils/team_utils.jsx';
+import * as Utils from 'utils/utils.jsx';
+
+import SystemPermissionGate from 'components/permissions_gates/system_permission_gate';
+import Pluggable from 'plugins/pluggable';
+
+import {ThreadsState} from 'mattermost-redux/types/threads';
+
+import TeamButton from './components/team_button';
+
+type Actions = {
+    getTeams: (page?: number, perPage?: number, includeTotalCount?: boolean) => void;
+    switchTeam: (url: string) => (dispatch: Dispatch<GenericAction>, getState: GetStateFunc) => void;
+    updateTeamsOrderForUser: (teamIds: string[]) => (dispatch: Dispatch<GenericAction>, getState: GetStateFunc) => Promise<void>;
+}
 
 type State = {
+    showOrder: boolean;
+    teamsOrder: Team[];
+}
+
+interface Props {
+    myTeams: Team[];
+    collapsedThreads: boolean;
+    currentTeamId: string;
+    match: { url: string };
+    moreTeamsToJoin: boolean;
+    myTeamMembers: Dictionary<TeamMembership>;
+    isOpen: boolean;
+    experimentalPrimaryTeam?: string ;
+    locale: string;
+    actions: Actions;
+    userTeamsOrderPreference: string;
+    threadCounts: ThreadsState['counts'];
+}
+
+export function renderView(props: Props) {
+    return (
+        <div
+            {...props}
+            className='scrollbar--view'
+        />);
+}
+
+export function renderThumbHorizontal(props: Props) {
+    return (
+        <div
+            {...props}
+            className='scrollbar--horizontal'
+        />);
+}
 
-};
+export function renderThumbVertical(props: Props) {
+    return (
+        <div
+            {...props}
+            className='scrollbar--vertical'
+        />);
+}
 
 export default class TeamSidebar extends React.PureComponent<Props, State> {
-    renderTeam = (team: Team) => {
-        return (
-            <TeamButton team={team}/>
+    constructor(props: Props) {
```

Comment:
This all came from replacing the dummy `TeamSidebar` component with all the code from the `LegacyTeamSidebar` that we never ended up re-writing. It's otherwise unchanged